### ISSUE TITLE
body_test_motion not reporting initial contact

### ISSUE THREAD

**Issue by godotproposer** [godotengine/godot#59126]
### Godot version

4.0.alpha3.official.256069eaf

### System information

Manjaro Linux

### Issue description

If a body is intersected, PhysicsServer2D.body_test_motion may report a different collider that the body isn't intersecting.
I suspect this ensures more consistent sliding behavior within CharacterBody2D.move_and_slide. But I think it's unfit for the scope of this function, making it unreliable for tests.

### Steps to reproduce

1. Position a body where it overlaps Collider A.
2. Position Collider B where it _would_ overlap the body after it was ejected from Collider A.
3. Write a script identifying which collider is reported by body_test_motion.
4. Observe as Collider B is reported first, although it did not initially overlap the body, whereas Collider A did. Note that subsequent reports are inconsistent.
(Or, open the minimal reproduction project, inspect the scene, then run & observe the console output.)

### Minimal reproduction project

[minimalrepro.zip](https://github.com/godotengine/godot/files/8241944/minimalrepro.zip)

**Comment by Zireael07** [comment]
Saw the same in 3D. Had to rely on Raycasts instead

**Comment by Chaosus** [comment]
Does it correctly reporting on 3.x branch?

**Comment by Zireael07** [comment]
(Not OP, but: seeing this on 3.4 release build)

**Comment by godotproposer** [comment]
> (Not OP, but: seeing this on 3.4 release build)

Same, I recreated the minimal repro in 3.4 with the same results

**Comment by Hiiamwilliam** [comment]
`body_test_motion` (on both 3.X and master) ends up calling an internal method `test_body_motion`. 

I don't have the courage to actually understand what's going on there, but if we are to trust the code comments, there are two that look important for this issue:

> //STEP 1, FREE BODY IF STUCK

then later

> // STEP 2 ATTEMPT MOTION

So maybe the fact that it's not reporting the initially intersected body could be intended? Who knows. However, if that's the case, it should be documented.

**Comment by godotproposer** [comment]
> So maybe the fact that it's not reporting the initially intersected body could be intended? Who knows. However, if that's the case, it should be documented.

I read the source and came to the same conclusion, yet I am convinced it is a design flaw. Currently, the method is unsuitable for tests, which is designated as its use case by name.

A more glaring flaw with the design is that the automation removes options from the user. If "step 1" of the method was removed, it would be trivial to replicate the current functionality of e.g. move_and_collide by calling it twice in the event of a collision. With "step 1" present, one cannot replicate the functionality where it isn't.

This is made worse because the lower-level physics code is not exposed to scripting. Thus, I cannot write my own variation of the method from script which lacks "step 1"; I would need to modify and build the engine from source. If those lower-level physics methods won't be exposed, then the exposed functions shouldn't assume what the user wants in this manner.

**Comment by AttackButton** [comment]
Yeah, got this issue very easily today trying to use `collide_separation_ray` param property. From long distance it confirms the test collision correctly. But up close these problems start to happen.

Move and collide use this function, should be a top priority issue.

**Comment by AttackButton** [comment]
Actually, it is a pretty neat code, just needs a tweak or two. Probably when the motion shape gets "stuck" on an object, it needs to set a mask to only collide with that object and then start the "free" loop. I'll try to hack it to see if it works.

And for `collide_separation_ray` param, an option when the object is stuck to only accept being "freed" by the closest output from the collider to the ray's origin, regardless of whether it's the longest output from the collider. Because sometimes the `collide separation ray` show the collision point as being on the other side of the collider. This could  also help with the ccd.
